### ISSUE TITLE
Runtime created objects via API stored in wrong path outside of config package (broken stage name)

### ISSUE THREAD

**Issue by dnsmichi** [Icinga/icinga2#7119]
## Describe the bug

Runtime created objects via the REST API live in the `_api` config package internally. If there are many concurrent requests, the `stageName` can be empty and the written config path is wrong.

Assume that the stageName is `abcd-1234`.

### Expected behaviour

Correct calculated in `ConfigObjectUtility::GetConfigDir()` as configDir.

```
packageDir = "/var/lib/icinga2/api/packages"
packageName = "_api"
activeStageName = "abcd-1234"
```

This information is used in `ConfigObjectUtility::GetObjectConfigPath(const Type::Ptr& type, const String& fullName)` to calculate the full config object path.

```
configDir = "/var/lib/icinga2/api/packages/_api/abcd-1234"

fullPath = configDir + "/conf.d/" + typeName + "/" + fullName + ".conf"
```

resulting in 

```
/var/lib/icinga2/api/packages/_api/abcd-1234/conf.d/host/mbpmif.int.netways.de.conf
```

### Broken behaviour

```
packageDir = "/var/lib/icinga2/api/packages"
packageName = "_api"
activeStageName = ""
```

resulting in

```
configDir = "/var/lib/icinga2/api/packages/_api/"

fullPath = configDir + "/conf.d/" + typeName + "/" + fullName + ".conf"
```

resulting in 

```
/var/lib/icinga2/api/packages/_api//conf.d/host/mbpmif.int.netways.de.conf
```

The activeStageName is read from the `/var/lib/icinga2/api/packages/_api/active-stage` file and with concurrent read/writes, this may call `fp.fail()` and return an empty string in case.

## To Reproduce

1. Create an object where the _api package is not yet created
2. Change permissions of `/var/lib/icinga2/api/packages/_api/active-stage` or delete its content
3. Create another object
4. See the changed path on disk where `conf.d` is located outside of the actual config path

## Additional Context

In order to forcefully create the error, create many objects in parallel, while reading the file lists from the package itself too.

### Config Packages explained in detail

Session with @Al2Klimov 

![icinga2_api_config_packages_session_AK jpg](https://user-images.githubusercontent.com/382049/56140332-fd40de80-5f9a-11e9-91ee-63b41dfc9215.jpg)

### Content of Config Packages affecting this bug

![icinga2_api_config_packages_cont_session_AK jpg](https://user-images.githubusercontent.com/382049/56140333-fd40de80-5f9a-11e9-844b-fa03e5c4ac3d.jpg)

## Possible Solution

- Store the knowledge of a package's active stage in memory and use the cached value

`GetActiveStage()` reads from the stored set m_ApiPackages[packageName]
`ActivateStage()` calls `SetActiveStage(packageName, stageName) and writes to m_ApiPackages[packageName] = stageName

- Better error handling with file IO and exceptions
- Additional safety logging when expected path handling is possibly broken
- Ensure that package files (include.conf, etc.) are intact - #6959 
- Figure out whether the StaticMutex is still needed when re-creating the API package

## Your Environment

Affects all versions since 2.4 with numerous attempted fixes in between.

## Issue Overview

FYI @lippserd 

- hosts added via API disappear after icinga restart #6580
- Host created via API, disappears after reload #4082
- Regenerate the _api/active-stage, _api/active.conf and _api/include.conf files when they're deleted #3668
- Downtime lost after restart #5625
- api/packages/_api/stagename/include.conf might get removed over restarts? #4797

**Comment by dnsmichi** [comment]
# Broken and validation error on startup

```
[2019-04-26 14:54:22 +0200] critical/ApiListener: Cannot detect active stage for package '_api'. Broken config package, check the troubleshooting documentation.
```

# Break the _api package

```
rm /usr/local/icinga/icinga2/var/lib/icinga2/api/packages/_api/active-stage
```

Deploy a new downtime.

```
curl -k -s -u root:icinga -H 'Accept: application/json' -X POST 'https://localhost:5665/v1/actions/schedule-downtime' -d '{ "start_time": 1556279961, "end_time": 1556289961, "duration": 1000, "author": "icingaadmin", "comment": "IPv4 network maintenance", "pretty": true, "type": "Service", "filter": true }'
```

`active-stage` is missing. But. The files are stored in the wrong directory.

```
$ ls -la /usr/local/icinga/icinga2/var/lib/icinga2/api/packages/_api/
total 16
drwx------  5 michi  wheel  160 Apr 26 13:56 .
drwx------  4 michi  wheel  128 Mar 27 14:39 ..
drwx------  5 michi  wheel  160 Feb  4 14:21 7e7861c8-8008-4e8d-9910-2a0bb26921bd
-rw-r--r--  1 michi  wheel  453 Feb  4 14:21 active.conf
-rw-r--r--  1 michi  wheel   25 Feb  4 14:21 include.conf
```

```
$ ls -la /usr/local/icinga/icinga2/var/lib/icinga2/api/packages/_api/7e7861c8-8008-4e8d-9910-2a0bb26921bd/conf.d/downtimes/
total 184
drwx------  25 michi  wheel  800 Apr 26 13:59 .
drwx------   3 michi  wheel   96 Feb  4 14:21 ..
-rw-r--r--   1 michi  wheel  412 Apr 26 13:59 many-test-0!many-test-0!6de085aa-bb2b-4395-b5f1-1f76a87ab636.conf
-rw-r--r--   1 michi  wheel  412 Apr 26 13:59 many-test-0!many-test-1!24af5654-4883-4dcc-8870-8cebb7e9a163.conf
-rw-r--r--   1 michi  wheel  412 Apr 26 13:59 many-test-0!many-test-2!afe3cd4c-2f60-4fdb-a92e-ffd9c558a483.conf
-rw-r--r--   1 michi  wheel  412 Apr 26 13:59 many-test-0!many-test-3!9386feb9-1238-427c-bcfb-033234226adb.conf
-rw-r--r--   1 michi  wheel  412 Apr 26 13:59 many-test-0!many-test-4!9e94f81f-4485-4a40-9a6b-218f366b69ad.conf
-rw-r--r--   1 michi  wheel  412 Apr 26 13:59 many-test-0!many-test-5!08227d24-211f-4c5c-88e1-bdc658a01fb7.conf
-rw-r--r--   1 michi  wheel  412 Apr 26 13:59 many-test-0!many-test-6!957f6166-65f8-425a-aca6-47e0895e5cee.conf
-rw-r--r--   1 michi  wheel  412 Apr 26 13:59 many-test-0!many-test-7!77135ca4-d921-4064-8068-695c9a6fa5a4.conf
-rw-r--r--   1 michi  wheel  412 Apr 26 13:59 many-test-0!many-test-8!ba2ba5ac-0d10-4f28-bae7-cef6ed255769.conf
-rw-r--r--   1 michi  wheel  412 Apr 26 13:59 many-test-0!many-test-9!c9f05c16-f31b-41b6-8ae9-8bdc1d977e6f.conf
-rw-r--r--   1 michi  wheel  417 Apr 26 13:59 mbpmif.int.netways.de!disk %2F!55936c99-2ac5-4f0e-993a-866b20c18d93.conf
-rw-r--r--   1 michi  wheel  415 Apr 26 13:59 mbpmif.int.netways.de!disk!3b285342-3ed7-401b-8cae-2905939ce7ec.conf
-rw-r--r--   1 michi  wheel  415 Apr 26 13:59 mbpmif.int.netways.de!http!1b5c61f6-3ba7-4bcf-ad43-4a9b15a14ca5.conf
-rw-r--r--   1 michi  wheel  417 Apr 26 13:59 mbpmif.int.netways.de!icinga!ebcf6ef6-c2d6-433a-9a48-700d7e1f945c.conf
-rw-r--r--   1 michi  wheel  532 Apr 26 04:00 mbpmif.int.netways.de!load!269febfc-b0ee-4420-8805-3b35a52ece6b.conf
-rw-r--r--   1 michi  wheel  415 Apr 26 13:59 mbpmif.int.netways.de!load!c27fd786-f029-4c58-bea3-ab2625e9c970.conf
-rw-r--r--   1 michi  wheel  416 Apr 26 13:59 mbpmif.int.netways.de!ping4!8e44c32c-b0f9-4149-a80a-5b9564d31f91.conf
-rw-r--r--   1 michi  wheel  416 Apr 26 13:59 mbpmif.int.netways.de!ping6!34e3729b-ef7b-470d-b488-9a716d7aaac7.conf
-rw-r--r--   1 michi  wheel  416 Apr 26 13:59 mbpmif.int.netways.de!procs!628a2409-9a6e-492c-a2e7-46b247c70e99.conf
-rw-r--r--   1 michi  wheel  414 Apr 26 13:59 mbpmif.int.netways.de!ssh!9124f047-18b9-40ab-aea8-d2fee327eb1d.conf
-rw-r--r--   1 michi  wheel  415 Apr 26 13:59 mbpmif.int.netways.de!swap!afcd3f64-7473-4fc4-b3a6-8eb7dd4423bc.conf
-rw-r--r--   1 michi  wheel  416 Apr 26 13:59 mbpmif.int.netways.de!users!74ba063a-096a-4b41-9f86-6fb6a7a326eb.conf
-rw-r--r--   1 michi  wheel  421 Apr 26 13:59 winmif!win-disk!ef93c368-253b-4182-bd2a-07b35dcfcde6.conf
```

On config validation, the API package will be broken.

```
$ icinga2 daemon -C
[2019-04-26 14:04:06 +0200] information/cli: Icinga application loader (version: v2.10.4-616-g921ee15fc; debug)
[2019-04-26 14:04:06 +0200] information/cli: Loading configuration file(s).
[2019-04-26 14:04:06 +0200] information/ConfigItem: Committing config item(s).
[2019-04-26 14:04:06 +0200] critical/ApiListener: Cannot detect active stage for package '_api'. Broken config package, check the troubleshooting documentation.
```

## Unbreak the _api package

With manual instructions, or by doing so from memory information. Unfortunately I cannot do this at startup where no cache exists ... that situation requires manual user interaction documented in the troubleshooting chapter.


Ensure it is fixed.
```
$ echo "7e7861c8-8008-4e8d-9910-2a0bb26921bd" > /usr/local/icinga/icinga2/var/lib/icinga2/api/packages/_api/active-stage
$ ls -la /usr/local/icinga/icinga2/var/lib/icinga2/api/packages/_api/active-stage
-rw-r--r--  1 michi  wheel  37 Apr 26 14:46 /usr/local/icinga/icinga2/var/lib/icinga2/api/packages/_api/active-stage
```

Start Icinga 2.

Break the package again at runtime.

michi@mbpmif ~/dev/icinga/icinga-vagrant/standalone (master=) $ rm /usr/local/icinga/icinga2/var/lib/icinga2/api/packages/_api/active-stage
```

Wait a bit, 60s interval for tests, live interval will be 5m.

```
[2019-04-26 14:47:26 +0200] information/ApiListener: Repairing broken API config package '_api', setting active stage '7e7861c8-8008-4e8d-9910-2a0bb26921bd'.
```

Verify it on disk.

```
michi@mbpmif ~/dev/icinga/icinga-vagrant/standalone (master=) $ ls -la /usr/local/icinga/icinga2/var/lib/icinga2/api/packages/_api/active-stage
-rw-r--r--  1 michi  wheel  36 Apr 26 14:47 /usr/local/icinga/icinga2/var/lib/icinga2/api/packages/_api/active-stage
michi@mbpmif ~/dev/icinga/icinga-vagrant/standalone (master=) $ cat /usr/local/icinga/icinga2/var/lib/icinga2/api/packages/_api/active-stage
7e7861c8-8008-4e8d-9910-2a0bb26921bd
```

Restart. Profit.

# Break Config Packages (Director etc.)

## Deploy a new stage

```
$ curl -k -s -u root:icinga -H 'Accept: application/json' -X POST -d '{"files": { "conf.d/test.conf": "object Host \"cmdb-host\" { check_command = \"dummy\" }" }, "pretty": true }' 'https://localhost:5665/v1/config/stages/d'
{
    "results": [
        {
            "code": 200.0,
            "package": "d",
            "stage": "c0bcc656-beb0-46c6-81d9-f702491e32d5",
            "status": "Created stage. Reload triggered."
        }
    ]
}
```

Verify everything is fine on disk.

```
$ cat /usr/local/icinga/icinga2/var/lib/icinga2/api/packages/d/active-stage
c0bcc656-beb0-46c6-81d9-f702491e32d5
```

## Remove active-stage and deploy a new stage

```
rm /usr/local/icinga/icinga2/var/lib/icinga2/api/packages/d/active-stage

$ curl -k -s -u root:icinga -H 'Accept: application/json' -X POST -d '{ "files": { "conf.d/test.conf": "object Host \"cmdb-host\" { check_command = \"dummy\" }" }, "pretty": true }' 'https://localhost:5665/v1/config/stages/d'
{
    "results": [
        {
            "code": 200.0,
            "package": "d",
            "stage": "1c0d8416-08f0-4e9e-a413-7d37315bc640",
            "status": "Created stage. Reload triggered."
        }
    ]
}
```

Works again.

```
cat /usr/local/icinga/icinga2/var/lib/icinga2/api/packages/d/active-stage
1c0d8416-08f0-4e9e-a413-7d37315bc640
```


# Docs

- Upgrading, since the new log message will appear
- Troubleshooting, linked from upgrading, including repair instructions

**Comment by dnsmichi** [comment]
include.conf/active.conf etc. are not part of this issue, with the fixed stage name always returning (or generating exceptions), these errors cannot happen anymore. Reviving these files is therefore not necessary. 

The main intent to solve:

- Runtime created objects exist after restart
- Reduce file I/O with active-stage reads, cache the details in memory
- Improve error handling, add exceptions and avoid wrong reads/writes
- Log config package errors on startup
- Repair missing active-stage files from a 5min timer, if they get deleted (brute force user action, the code uses lock guards now).

**Comment by ekeih** [comment]
@dnsmichi This fix is hihgly appreciated! I am looking forward to 2.11.0 🚀 
Thank you!
